// wordreference/lib/dictionaries.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.wrDictionaryLookup = void 0;
exports.wrDictionaryLookup = {
    enar: 'English-Arabic',
    enzh: 'English-Chinese',
    encz: 'English-Czech',
    ennl: 'English-Dutch',
    enfr: 'English-French',
    ende: 'English-German',
    engr: 'English-Greek',
    enis: 'English-Icelandic',
    enit: 'English-Italian',
    enja: 'English-Japanese',
    enko: 'English-Korean',
    enpl: 'English-Polish',
    enpt: 'English-Portuguese',
    enro: 'English-Romanian',
    enru: 'English-Russian',
    enes: 'English-Spanish',
    ensv: 'English-Swedish',
    entr: 'English-Turkish',
    aren: 'Arabic-English',
    czen: 'Czech-English',
    deen: 'German-English',
    dees: 'German-Spanish',
    esde: 'Spanish-German',
    esen: 'Spanish-English',
    esfr: 'Spanish-French',
    esit: 'Spanish-Italian',
    espt: 'Spanish-Portuguese',
    fren: 'French-English',
    fres: 'French-Spanish',
    gren: 'Greek-English',
    isen: 'Icelandic-English',
    iten: 'Italian-English',
    ites: 'Italian-Spanish',
    jaen: 'Japanese-English',
    koen: 'Korean-English',
    nlen: 'Dutch-English',
    plen: 'Polish-English',
    pten: 'Portuguese-English',
    ptes: 'Portuguese-Spanish',
    roen: 'Romanian-English',
    ruen: 'Russian-English',
    sven: 'Swedish-English',
    tren: 'Turkish-English',
    zhen: 'Chinese-English',
};
